crates/engine/invalid-block-hooks/src/witness.rs

```diff
@@ -1,13 +1,247 @@
+            evm.db_mut().commit(result.state);
+        }
+
+        drop(evm);
+
+        // Merge all state transitions
+        db.merge_transitions(BundleRetention::Reverts);
+
+        // Take the bundle state
+        let bundle_state = db.take_bundle();
+
+        // Initialize a map of preimages.
+        let mut state_preimages = HashMap::new();
+
+        // Grab all account proofs for the data accessed during block execution.
+        //
+        // Note: We grab *all* accounts in the cache here, as the `BundleState` prunes
+        // referenced accounts + storage slots.
+        let mut hashed_state = HashedPostState::from_bundle_state(&bundle_state.state);
+        for (address, account) in db.cache.accounts {
+            let hashed_address = keccak256(address);
+            hashed_state
+                .accounts
+                .insert(hashed_address, account.account.as_ref().map(|a| a.info.clone().into()));
+
+            let storage = hashed_state
+                .storages
+                .entry(hashed_address)
+                .or_insert_with(|| HashedStorage::new(account.status.was_destroyed()));
+
+            if let Some(account) = account.account {
+                state_preimages.insert(hashed_address, alloy_rlp::encode(address).into());
+
+                for (slot, value) in account.storage {
+                    let slot = B256::from(slot);
+                    let hashed_slot = keccak256(slot);
+                    storage.storage.insert(hashed_slot, value);
+
+                    state_preimages.insert(hashed_slot, alloy_rlp::encode(slot).into());
+                }
+            }
+        }
+
+        // Generate an execution witness for the aggregated state of accessed accounts.
+        // Destruct the cache database to retrieve the state provider.
+        let state_provider = db.database.into_inner();
+        let witness = state_provider.witness(HashedPostState::default(), hashed_state.clone())?;
+
+        // Write the witness to the output directory.
+        let mut file = File::options()
+            .write(true)
+            .create_new(true)
+            .open(self.output_directory.join(format!("{}_{}.json", block.number, block.hash())))?;
+        let response = ExecutionWitness { witness, state_preimages: Some(state_preimages) };
+        file.write_all(serde_json::to_string(&response)?.as_bytes())?;
+
+        // The bundle state after re-execution should match the original one.
+        if let Some(path) = (bundle_state != output.state)
+            .then(|| {
+                self.save_diff(
+                    format!("{}_{}.bundle_state.diff", block.number, block.hash()),
+                    &bundle_state,
+                    &output.state,
+                )
+            })
+            .transpose()?
+        {
+            warn!(target: "engine::invalid_block_hooks::witness", path = %path.display(), "Bundle state mismatch after re-execution");
+        }
```

Comment:
can we just do pls 🙏 
```suggestion
        if bundle_state != output.state {
            let filename = format!("{}_{}.bundle_state.diff", block.number, block.hash());
            let path = self.save_diff(filename, &bundle_state, &output.state)?;
            warn!(target: "engine::invalid_block_hooks::witness", path = %path.display(), "Bundle state mismatch after re-execution");
        }
        if let Some(path) = (bundle_state != output.state)
            .then(|| {
                self.save_diff(
                    format!("{}_{}.bundle_state.diff", block.number, block.hash()),
                    &bundle_state,
                    &output.state,
                )
            })
            .transpose()?
        {
            warn!(target: "engine::invalid_block_hooks::witness", path = %path.display(), "Bundle state mismatch after re-execution");
        }
```